### ISSUE TITLE
`podman build --squash` removes intermediate image cached layers

### ISSUE THREAD

**Issue by mtalexan** [containers/podman#20824]
### Issue Description

When building with `docker buildx build --squash ...` intermediate layers are retained for use as a build layer cache.  But `podman build --squash ...` removes the intermediate layers during squashing, effectively adding `--no-cache`.

### Steps to reproduce the issue

Steps to reproduce the issue
1. Create a simple multi-command Containerfile
```dockerfile
FROM fedora:39

WORKDIR /app

RUN dnf update -y
RUN dnf install -y jq
RUN dnf clean all
```
2. Build it twice
```shell
podman build --squash --progress=plain -t myimage -f Containerfile .
```

### Describe the results you received

The second time it builds, no layers are pulled from the cache.  Furthermore, the intermediate layers are not present in the cache after any number of builds with `--squash` included.

### Describe the results you expected

The second time it builds, all the layers should still be present in the cache and used from there.  Output for the second build should look something like what docker produces:
```
#1 [internal] load build definition from Containerfile
#1 transferring dockerfile: 128B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/library/fedora:39
#3 DONE 0.4s

#4 [1/5] FROM docker.io/library/fedora:39@sha256:ee16ca86648f857c68270514442953c808c61c416152b9a316e551b02bedf8ec
#4 DONE 0.0s

#5 [4/5] RUN dnf install -y jq
#5 CACHED

#6 [2/5] WORKDIR /app
#6 CACHED

#7 [3/5] RUN dnf update -y
#7 CACHED

#8 [5/5] RUN dnf clean all
#8 CACHED

#9 exporting to image
#9 exporting layers done
#9 writing image sha256:0845bd6118bc97cdd89c3a1d58e4fb1b66937d3f3802721c8e3f21138d8cf611 done
#9 naming to docker.io/library/myimage done
#9 DONE 0.0s
```

And intermediate layers should be present in the cache after every build, regardless of whether or not `--squash` is included.

### podman info output

```yaml
TL;DR
Host: Ubuntu 22.04 amd64  
Podman: 4.7.0 


host:
  arch: amd64
  buildahVersion: 1.32.0
  cgroupControllers:
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon_2.0.25+ds1-1.1_amd64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.25, commit: unknown'
  cpuUtilization:
    idlePercent: 95.77
    systemPercent: 1.01
    userPercent: 3.22
  cpus: 12
  databaseBackend: boltdb
  distribution:
    codename: jammy
    distribution: ubuntu
    version: "22.04"
  eventLogger: journald
  freeLocks: 1986
  hostname: goln-422q533
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1684400513
      size: 1
    - container_id: 1
      host_id: 231072
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1684455083
      size: 1
    - container_id: 1
      host_id: 231072
      size: 65536
  kernel: 6.2.0-37-generic
  linkmode: dynamic
  logDriver: journald
  memFree: 13639499776
  memTotal: 67143614464
  networkBackend: cni
  networkBackendInfo:
    backend: cni
    dns:
      package: golang-github-containernetworking-plugin-dnsname_1.3.1+ds1-2_amd64
      path: /usr/lib/cni/dnsname
      version: |-
        CNI dnsname plugin
        version: 1.3.1
        commit: unknown
    package: 'containernetworking-plugins, golang-github-containernetworking-plugin-dnsname:
      /usr/lib/cni'
    path: /usr/lib/cni
  ociRuntime:
    name: crun
    package: crun_0.17+dfsg-1.1_amd64
    path: /usr/bin/crun
    version: |-
      crun version 0.17
      commit: 0e9229ae34caaebcb86f1fde18de3acaf18c6d9a
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  pasta:
    executable: ""
    package: ""
    version: ""
  remoteSocket:
    exists: true
    path: /run/user/1684455083/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /nix/store/sib9d0gq6vm33kbjgy2q713ncg3qv0mh-podman-4.7.0/libexec/podman/slirp4netns
    package: Unknown
    version: |-
      slirp4netns version 1.2.2
      commit: 0ee2d87523e906518d34a6b423271e4826f71faf
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.4
  swapFree: 1218957312
  swapTotal: 1218957312
  uptime: 33h 51m 57.00s (Approximately 1.38 days)
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries:
  search:
  - docker.io
  - quay.io
store:
  configFile: /home/mtalexander/.config/containers/storage.conf
  containerStore:
    number: 2
    paused: 0
    running: 2
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/mtalexander/.local/share/containers/storage
  graphRootAllocated: 503336263680
  graphRootUsed: 216475414528
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Supports shifting: "false"
    Supports volatile: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 73
  runRoot: /run/user/1684455083/containers
  transientStore: false
  volumePath: /home/mtalexander/.local/share/containers/storage/volumes
version:
  APIVersion: 4.7.0
  Built: 315532800
  BuiltTime: Mon Dec 31 19:00:00 1979
  GitCommit: ""
  GoVersion: go1.21.3
  Os: linux
  OsArch: linux/amd64
  Version: 4.7.0
```
```


### Podman in a container

No

### Privileged Or Rootless

None

### Upstream Latest Release

No

### Additional environment details

Podman is installed via `nix`, which is why the paths to the executables look strange, but the issue is fully reproducible with podman-in-podman, podman-in-docker, and on a Fedora VM with native podman installed.

The latest version of Podman (4.8.0) isn't yet packaged by upstream yet on any system I have an easy way to use, which is why the latest hasn't been tested.

### Additional information

There have been a number of issues like this reported before, but there's always confusion about `--squash` vs `--squash-all`.  The `--squash` option is one `docker` also provides and is declared to retain the build layer cache so subsequent builds can re-use the layers.  To retain compatibility with docker, podman should do the same and this issue is reporting that it doesn't.    
The `--squash-all` option is a podman/buildah-unique option and currently requires `--layers` to be used with it if the intermediate layers should be retained.  This is fine (though probably sub-optimal) since there's no compatibility with docker that needs to be met.  
Notably, `--layers` cannot be used with `--squash` in podman, it triggers a CLI error.  Though requiring that extra option, assuming the error from the CLI preventing the combination was removed, would still leave podman incompatible with the docker CLI option.

**Comment by flouthoc** [comment]
This needs to be fixed at buildah end, I'll check this although this has been like this since 2018 https://github.com/containers/buildah/issues/1234 

The problem here is that `podman build --squash` is different from `podman build --squash-all` See: https://docs.podman.io/en/stable/markdown/podman-build.1.html, I think `podman build --sqaush` is just `buildah build --layers=false` and this is how it was since beginning See: https://github.com/containers/buildah/issues/1234

Buildah needs a new option so it can create layers but only squash newly created layers and old layers are intact this is different from what `buildah` does with `--squash` where it squashes all of the layers including per-exisiting ones.

I'll check this.

**Comment by flouthoc** [comment]
For now I'll suggest you can use `--squash-all` with `--layers` which should temporarily unblock you.

**Comment by mtalexan** [comment]
> This needs to be fixed at buildah end, I'll check this although this has been like this since 2018 [containers/buildah#1234](https://github.com/containers/buildah/issues/1234)

It appears `docker buildx` had this optionnsincenaround that time too, though it preserved the cache layers so they could be used during a rebuild/other builds.

> Buildah needs a new option so it can create layers but only squash newly created layers and old layers are intact this is different from what `buildah` does with `--squash` where it squashes all of the layers including per-exisiting ones.

I was afraid that's what was being done. That's a big ask to add logic for squashing at the end rather than never creating the separate layers in the first place. Unfortunately it's a docker compatibility thing though, and also has huge impact on multi-stage builds (subsequent stages rely on the build cache to provide the intermediate layers for the prior stage when using one as a base image).

> For now I'll suggest you can use `--squash-all` with `--layers` which should temporarily unblock you.

Unfortunately that's not equivalent. That loses the base image layer by squashing it as well, which means the base images can't be shared at deployment.

Technically the only equivalent is to mentally track every change made to your target stage, make them in a sub-stage instead, create a new stage from the same original base image, and then do one big `RUN --mount` to copy every changed thing into a new final "squashed" stage at once. This of course assumes you don't have any `USER`, `VOLUME`, `ENTRYPOINT`, `ENV`, `CMD`, etc which can't be replicated without creating a new layer, but do get squashed.

**Comment by flouthoc** [comment]
I think this is similar to this https://github.com/containers/buildah/issues/6257

**Comment by victorserbu2709** [comment]
buildah seems fixed: https://github.com/podman-container-tools/buildah/issues/6257
